Send milliseconds in timestamp to Logstash. Reduce dependencies. Add TCP (+optional SSL)

derekkraan opened the pull request:
I installed this yesterday and noticed that the milliseconds were missing from all my timestamps. This PR fixes that issue. I've also modernized the date / time handling code and made the JSON encoder module configurable (also changed the default to `Jason`).

I've updated the TravisCI config to remove Elixir 1.3.0 from the build targets.

derekkraan committed:
Send milliseconds in timestamp to Logstash. Remove dependency on `Timex`
and `exjsx`.

derekkraan committed:
Update TravisCI build targets

meatherly commented on README.md:
```diff
@@ -16,6 +16,8 @@ that will send logs to the [Logstash UDP input](https://www.elastic.co/guide/en/
  * **level**: Atom. Minimum level for this backend.
  * **type**: String.t. Type of logs. Useful to filter in logstash.
+ * **timezone**: String.t. Server timezone. Used to convert from naive timestamp. Default `"Etc/UTC"`.
+ * **json_encoder**: Atom. Module to be used for JSON encoding. Default `Jason`.
```
Since `Jason` is only in the `:dev` environment, should we add a note to include it in dependency list if they haven't already?

derekkraan replied:
Good point. I've also removed it from the `dev` dependencies (so now it's only in `test`) so that nobody deploys a broken system to production by accident.

meatherly commented on mix.lock:
```diff
@@ -1,14 +1,17 @@
-%{"certifi": {:hex, :certifi, "0.7.0", "861a57f3808f7eb0c2d1802afeaae0fa5de813b0df0979153cbafcd853ababaf", [:rebar3], []},
```
If you run `mix do deps.unlock --unused, deps.clean --unused` it should clean this file up a lot. I don't know if it matters having unused deps in the mix.lock file or not though.

derekkraan committed:
Clean up deps & add note to install `Jason` to use it.

derekkraan committed:
Allow configuring protocol and add :tcp as option

derekkraan committed:
Add support for TLS connections

derekkraan committed:
Remove unused variables

derekkraan committed:
Pattern match output of :ssl.connect/2

derekkraan committed:
Match on output of `send` functions

derekkraan committed:
Add fields to top level

derekkraan committed:
Merge pull request #1 from derekkraan/tcp

Tcp

derekkraan commented:
I've since updated my `master` branch to include support for TCP (with and without TLS) protocol (loosely based on #8)

derekkraan committed:
Update tests to reflect fields being merged directly into json

derekkraan committed:
Handle unavailable connections by attempting reconnect on every log

derekkraan committed:
Only log error once

derekkraan committed:
Log as error

derekkraan committed:
Merge pull request #2 from derekkraan/graceful_tcp_not_available

Graceful tcp not available

derekkraan commented:
And now also updated to gracefully handle not being able to reach logstash (so your whole app doesn't crash if logstash is down for whatever reason).

derekkraan committed:
Reset recorded_error when connection has been made

derekkraan committed:
Set a connection timeout of 10s and a send timeout of 10s.

The default value for connection timeout is infinity, in my case it was
waiting for an hour before giving up, which is an awfully long time.

derekkraan commented:
I've updated this further to add some timeouts. We were having issues where a new deploy of our app wasn't detecting a closed connection for an hour before it crashed and re-opened the connection. That should happen much quicker now.

derekkraan commented:
Hi @marcelog, I've just been adding to and improving this PR as I go along, any chance you'll be able to find some time to review it soon?

derekkraan committed:
Handle closed connections (log error and mark socket as closed)

derekkraan committed:
Specify timeout for :ssl.connect (and reduce other timeouts)
